File: perl/ch01/p1_1.pl
#!/usr/bin/perl

# Page 20, problem 1.1.

use strict;
use warnings;

# Simple solution using division
sub array_prods_simple {
    my @array = @_;

    my $full = 1;
    for my $i (@array) {
        $full *= $i;
    }

    return (map { $full / $_ } @array);
}

# Solution without division
sub array_prods {
    my @array = @_;
    my $last  = $#array;

    # Compute partials
    my @forward  = ();
    my @backward = ();

    for my $i (0 .. $last) {
        if ($i == 0) {
            $forward[0]      = $array[0];
            $backward[$last] = $array[$last];
        } else {
            $forward[$i] = $array[$i] * $forward[$i - 1];
            $backward[$last - $i] =
                $array[$last - $i] * $backward[$last - $i + 1];
        }
    }

    my @prods = ();
    for my $i (0 .. $last) {
        if ($i == 0) {
            push @prods, $backward[$i + 1];
        } elsif ($i == $last) {
            push @prods, $forward[$i - 1];
        } else {
            push @prods, $forward[$i - 1] * $backward[$i + 1];
        }
    }

    return @prods;
}

if (@ARGV) {
    my @a1 = array_prods_simple(@ARGV);
    my @a2 = array_prods(@ARGV);
    print "\@a1=(@a1)\n";
    print "\@a2=(@a2)\n";
} else {
    my @l = qw(1 2 3 4 5);
    my @a1 = array_prods_simple(@l);
    my @a2 = array_prods(@l);
    print "\@a1=(@a1)\n";
    print "\@a2=(@a2)\n";
}

File: java/ch01/p1_1.java
/*
    Page 20, problem 1.1.
 */

import java.util.Arrays;

public class p1_1 {
    /*
        Simple solution, using division
     */
    private static long[] arrayProdsSimple(long[] list) {
        long[] prods = new long[list.length];
        long allProd = 1;

        for (long x : list) {
            allProd *= x;
        }

        for (int i = 0; i < list.length; i++) {
            prods[i] = allProd / list[i];
        }

        return prods;
    }

    /*
        Non-division solution.
     */
    private static long[] arrayProds(long[] list) {
        long[] prods = new long[list.length];
        long[] forward = new long[list.length];
        long[] backward = new long[list.length];
        int size = list.length;

        // Compute the partials
        for (int i = 0; i < size; i++) {
            if (i == 0) {
                forward[i] = list[i];
                backward[size - 1] = list[size - 1];
            } else {
                forward[i] = list[i] * forward[i - 1];
                backward[size - i - 1] = list[size - i - 1] * backward[size - i];
            }
        }

        // Compute the products
        for (int i = 0; i < size; i++) {
            if (i == 0) {
                prods[i] = backward[1];
            } else if (i == (size - 1)) {
                prods[i] = forward[i - 1];
            } else {
                prods[i] = forward[i - 1] * backward[i + 1];
            }
        }

        return prods;
    }

    public static void main(String[] args) {
        long[] list;

        if (args.length > 0) {
            list = new long[args.length];
            for (int i = 0; i < args.length; i++) {
                list[i] = Long.parseLong(args[i]);
            }
        } else {
            list = new long[]{1L, 2L, 3L, 4L, 5L};
        }

        System.out.println(Arrays.toString(arrayProdsSimple(list)));
        System.out.println(Arrays.toString(arrayProds(list)));
    }
}